In [1]:
import numpy as np
from scipy import stats

# Семинар 9

## Задача1

Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks): zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110], ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832]. Используя математические операции, посчитать коэффициенты линейной регрессии, приняв за X заработную плату (то есть, zp - признак), а за y - значения скорингового балла (то есть, ks - целевая переменная). Произвести расчет как с использованием intercept, так и без.

In [2]:
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])

In [3]:
b1 = (np.mean(zp*ks)-np.mean(ks)*np.mean(zp))/(np.mean(zp**2)-np.mean(zp)**2)
b1

2.620538882402765

In [4]:
b0 = np.mean(ks)-b1*np.mean(zp)
b0

444.1773573243596

In [5]:
y_pred = b0+b1*zp
y_pred

array([535.89621821, 562.10160703, 942.07974498, 968.2851338 ,
       548.99891262, 627.61507909, 585.68645697, 837.25818968,
       758.64202321, 732.43663439])

In [6]:
# расчет без intercept
b_wi = np.sum(zp * ks) / np.sum(zp ** 2)
b_wi

5.889820420132689

In [7]:
y_pred = b_wi*zp
y_pred

array([ 206.1437147 ,  265.04191891, 1119.06587983, 1177.96408403,
        235.59281681,  412.28742941,  318.05030269,  883.47306302,
        706.77845042,  647.88024621])

## Задача 2

Посчитать коэффициент линейной регрессии при заработной плате (zp), используя градиентный спуск (без intercept).

In [8]:
x = zp
y = ks
x,y

(array([ 35,  45, 190, 200,  40,  70,  54, 150, 120, 110]),
 array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832]))

In [9]:
def mse_ (B1,y=y, x=x,n =10):
    return np.sum((B1*x-y)**2)/n

In [10]:
alpha = 1e-6

In [11]:
B1 = 0.1
n = 10

In [12]:
for i in range(30000):
    B1-= alpha *(2/n)* np.sum((B1*x-y)*x)
    if i % 500 == 0:
        print("Iteration={i} B1={B1} mse={mse}". format(i=i, B1=B1, mse=mse_(B1)))

Iteration=0 B1=0.25952808 mse=493237.7212546963
Iteration=500 B1=5.889815595583751 mse=56516.858416040064
Iteration=1000 B1=5.8898204201285544 mse=56516.85841571941
Iteration=1500 B1=5.889820420132673 mse=56516.85841571943
Iteration=2000 B1=5.889820420132673 mse=56516.85841571943
Iteration=2500 B1=5.889820420132673 mse=56516.85841571943
Iteration=3000 B1=5.889820420132673 mse=56516.85841571943
Iteration=3500 B1=5.889820420132673 mse=56516.85841571943
Iteration=4000 B1=5.889820420132673 mse=56516.85841571943
Iteration=4500 B1=5.889820420132673 mse=56516.85841571943
Iteration=5000 B1=5.889820420132673 mse=56516.85841571943
Iteration=5500 B1=5.889820420132673 mse=56516.85841571943
Iteration=6000 B1=5.889820420132673 mse=56516.85841571943
Iteration=6500 B1=5.889820420132673 mse=56516.85841571943
Iteration=7000 B1=5.889820420132673 mse=56516.85841571943
Iteration=7500 B1=5.889820420132673 mse=56516.85841571943
Iteration=8000 B1=5.889820420132673 mse=56516.85841571943
Iteration=8500 B1=5.889

Коэффициент линейной регрессии без interceptor = 5.8898

    ## Задача 3
Произвести вычисления как в пункте 2, но с вычислением intercept. Учесть, что изменение коэффициентов должно производиться на каждом шаге одновременно (то есть изменение одного коэффициента не должно влиять на изменение другого во время одной итерации).

In [13]:
def mse0_ (B1, B0,y=y, x=x,n =10):
     return np.sum((B0 + B1*x-y)**2)/n

In [14]:
alpha = 1e-5
B0 = 0.1
B1 = 0.1
for i in range(3000001):
    B0-= alpha *(2/n)* np.sum((B0+B1*x-y))

    B1-= alpha *(2/n)* np.sum((B0+B1*x-y)*x)
    if i % 500000 == 0:
        print("Iteration={i} B1={B1}  B0={B0} mse={mse}". format(i=i, B1=B1,B0=B0, mse=mse0_(B1,B0)))

Iteration=0 B1=1.6950496217904003  B0=0.1139932 mse=298808.66922912706
Iteration=500000 B1=2.8791839426649335  B0=409.03729808290353 mse=6783.645660253098
Iteration=1000000 B1=2.6410076607593775  B0=441.3964262548758 mse=6472.375938447427
Iteration=1500000 B1=2.622158750522415  B0=443.9572786511766 mse=6470.426487340797
Iteration=2000000 B1=2.6206670763125217  B0=444.15994063146525 mse=6470.414278123679
Iteration=2500000 B1=2.6205490274745853  B0=444.1759789936681 mse=6470.414201658573
Iteration=3000000 B1=2.620539685268393  B0=444.177248245355 mse=6470.414201179681
